### PR TITLE
Redesign workflow engine and provisioner API

### PR DESCRIPTION
**Pull request by apmasell**
This replaces the call-back based `WorkMonitor` with the fluent monad `OperationAction` API. This change will prevent recovery of any inflight operations.

Jira ticket:

- [X] Includes a change file
- [X] Updates developer documentation

### CODE REVIEW

**Comment by avarsava** [comment]
> This change will prevent recovery of any inflight operations.

Can you elaborate on this?

**Comment by apmasell** [comment]
The API change will affect what recovery information is stored in the database, so any workflow that attempted to be recovered when this feature is first deployed will fail.

**Comment by avarsava** [comment]
OK, but after that recovery will work?

**Comment by apmasell** [comment]
Yup. Maybe I should include a migration that drops any running workflows.

File: plugin-guide.md [hunk 1]
```diff
@@ -116,9 +116,7 @@ shared directory instead of, say, `/tmp` and ensure the right permissions are
 set up.
 These are not the responsibility of the plugin author.
 
-The class `BaseJsonInputProvisioner` is a partial implementation that can store
-crash recovery information in a JSON object of the implementor's choosing,
-making recovery easier.
+This plugin type uses the operations API.
```
Comment (avarsava) [inline plugin-guide.md:119]: Can these be a link to the section? if markdown can do that?

File: plugin-guide.md [hunk 2]
```diff
@@ -211,6 +203,66 @@ to filter on runs. A filter could query Pinery and get all the external
...
+previous step is the only parameter to the next one. This design is preferable
+to direct calls because Víðarr can stop executing the task when it needs to
+wait and restart it later, removing the burden of asynchronous scheduling from
+the plugin author.
+
+To start, a call to `OperationAction.load` or `OperationAction.value` is
+required. This primes the sequence of events by computing a value from the
+state information alone which will be the input for the first step. After this,
+`then` may be called to manipulate this value. Additionally, there are
+convinced methods `map`, to modify the current value, and `reload`, to discard
```
Comment (avarsava) [inline plugin-guide.md:238]: "convinced" methods?

Comment (apmasell) [reply plugin-guide.md]: convenience. Fixed.

File: plugin-guide.md (hunk 1 above)
Comment (apmasell) [reply plugin-guide.md]: Done.

File: plugin-guide.md [hunk 3]
```diff
@@ -174,17 +168,15 @@ which ones are allowed via the `supports` method.
 The workflow engine will be given the complete input to the workflow (with real
 paths provided by the input provisioners) and the workflow itself. Once the
 workflow has completed, it must provide a JSON structure that references the
-output of the workflow. Vidarr will identified the output files generated by
+output of the workflow. Víðarr will identified the output files generated by
```
Comment (avarsava) [inline plugin-guide.md:171]: "will identified"

Comment (apmasell) [reply plugin-guide.md]: Fixed.

File: vidarr-cromwell/src/main/java/ca/on/oicr/gsi/vidarr/cromwell/CromwellOutputProvisioner.java [hunk 4]
```diff
@@ -314,149 +173,62 @@ public String getWorkflowUrl() {
...
+                    String.format("Got response %d on %s", response.statusCode(), cromwellUrl)))
+        .then(monitorWhen(CROMWELL_FAILURES, result -> result.statusCode() / 100 != 2, cromwellUrl))
+        .then(requireJsonSuccess())
+        .map(result -> Optional.ofNullable(result.getId()).filter(id -> !id.equals("null")))
+        .then(requirePresent())
+        .then(status(WorkingStatus.QUEUED))
+        .then(
+            log(
+                Level.INFO,
+                id -> String.format("Started Cromwell workflow %s on %s", id, cromwellUrl)))
```
[the remark this answers is not in the document]
Comment (apmasell) [reply vidarr-cromwell/src/main/java/ca/on/oicr/gsi/vidarr/cromwell/CromwellOutputProvisioner.java:202]: Mostly due to copy-paste. I'll fix that. Also, the logs get prefixed with context, so it wouldn't be hard to tell.

**Comment by apmasell** [comment]
Also, the build issue seems to be a rebasing problem.

File: vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/OperationAction.java [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        return MAPPER.valueToTree(s);
+      }
+    };
+  }
+
+  /**
+   * Recover and operation from the database and discard and error conditions and allow it to start
+   * again.
+   *
+   * @param state the
```
Comment (avarsava) [inline vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/OperationAction.java:273]: the?

Comment (apmasell) [reply vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/OperationAction.java]: Drop the _the_. It's cleaner.

File: vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/WorkflowEngine.java [hunk 6]
```diff
@@ -24,9 +26,31 @@
 /** Defines an engine that knows how to execute workflows and track the results */
 @JsonTypeIdResolver(WorkflowEngine.WorkflowEngineIdResolver.class)
 @JsonTypeInfo(use = JsonTypeInfo.Id.CUSTOM, include = As.PROPERTY, property = "type")
-public interface WorkflowEngine {
+public interface WorkflowEngine<S extends Record, C extends Record> {
```
Comment (avarsava) [inline vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/WorkflowEngine.java:29]: What are S and C here?

Comment (apmasell) [reply vidarr-pluginapi/src/main/java/ca/on/oicr/gsi/vidarr/WorkflowEngine.java]: Renamed for clarity and added JavaDoc.

**Review by avarsava — APPROVED** [review]

File: vidarr-server/src/main/resources/db/migration/V0019__delete_recovery.sql [hunk 7]
```diff
@@ -0,0 +1,6 @@
+DELETE FROM active_operation WHERE workflow_run_id IN (SELECT active_workflow_run.id FROM active_workflow_run);
+DELETE FROM analysis_external_id WHERE analysis_external_id.external_id_id IN (SELECT external_id.id FROM external_id WHERE external_id.workflow_run_id IN (SELECT active_workflow_run.id FROM active_workflow_run));
+DELETE FROM external_id WHERE external_id.workflow_run_id IN (SELECT active_workflow_run.id FROM active_workflow_run);
```
Comment (avarsava) [inline vidarr-server/src/main/resources/db/migration/V0019__delete_recovery.sql:3]: ERROR: update or delete on table "external_id" violates foreign key constraint "external_id_version_external_id_id_fkey" on table "external_id_version"

Comment (apmasell) [reply vidarr-server/src/main/resources/db/migration/V0019__delete_recovery.sql:4]: Should be fixed.

**Review by grapigeau — APPROVED** [review]

**Pull request merged**